lib/Doctrine/ORM/PersistentCollection.php

```diff
@@ -231,13 +232,12 @@ public function getSnapshot(): array
      */
     public function getDeleteDiff(): array
     {
-        return array_udiff_assoc(
-            $this->snapshot,
-            $this->collection->toArray(),
-            static function ($a, $b): int {
-                return $a === $b ? 0 : 1;
-            }
-        );
+        $collectionItems = $this->collection->toArray();
+
+        return array_values(array_diff_key(
+            array_combine(array_map('spl_object_id', $this->snapshot), $this->snapshot),
+            array_combine(array_map('spl_object_id', $collectionItems), $collectionItems)
```

Comment:
`spl_object_id` is not available on 7.1, but https://github.com/doctrine/orm/pull/8837 will introduce a polyfill for it, so let's merge that PR first.